src/Common/src/System/Security/Cryptography/AsnReader.cs

```diff
@@ -0,0 +1,2806 @@
+                    }
+
+                    charsWritten = encoding.GetChars(bytePtr, source.Length, charPtr, destination.Length);
+                    Debug.Assert(charCount == charsWritten);
+                }
+                catch (DecoderFallbackException e)
+                {
+                    throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding, e);
+                }
+
+                return true;
+            }
+        }
+
+        private delegate bool CharacterStringProcessor(
+            ReadOnlyMemory<byte> data,
+            Span<char> dest,
+            Text.Encoding encoding);
+
+        private bool ProcessConstructedCharacterString(
+            ReadOnlyMemory<byte> source,
+            Span<char> dest,
+            bool isIndefinite,
+            Text.Encoding encoding,
+            CharacterStringProcessor processor)
+        {
+            int bytesLength = CountConstructedOctetString(source, isIndefinite);
+
+            byte[] rented = ArrayPool<byte>.Shared.Rent(bytesLength);
+
+            try
+            {
+                // T-REC-X.690-201508 sec 8.23.3, all character strings are encoded as octet strings.
+                CopyConstructedOctetString(source, rented, isIndefinite, out int bytesRead, out int bytesWritten);
+                Debug.Assert(bytesWritten == bytesLength);
+
+                bool processed = processor(
+                    new ReadOnlyMemory<byte>(rented, 0, bytesWritten),
+                    dest,
+                    encoding);
+
+                if (processed)
+                {
+                    _data = _data.Slice(bytesRead);
+                }
+
+                return processed;
+            }
+            finally
+            {
+                Array.Clear(rented, 0, bytesLength);
+                ArrayPool<byte>.Shared.Return(rented);
+            }
+        }
+
+        private bool ProcessCharacterString(
+            Asn1Tag expectedTag,
+            UniversalTagNumber universalTagNumber,
+            Text.Encoding encoding,
+            Span<char> dest,
+            CharacterStringProcessor processor)
+        {
+            // T-REC-X.690-201508 sec 8.23.3, all character strings are encoded as octet strings.
+
+            if (TryGetPrimitiveOctetStringBytes(
+                expectedTag,
+                out Asn1Tag actualTag,
+                out int? contentLength,
+                out int headerLength,
+                out ReadOnlyMemory<byte> contents,
+                universalTagNumber))
+            {
+                bool processed = processor(contents, dest, encoding);
+
+                if (processed)
+                {
+                    _data = _data.Slice(headerLength + contents.Length);
+                }
+
+                return processed;
+            }
+
+            Debug.Assert(actualTag.IsConstructed);
+
+            return ProcessConstructedCharacterString(
+                Slice(_data, headerLength, contentLeng
```

Comment:
You have lots of cases with the TryGetPrimitive/TryCopyConstructed pattern. Is there a way to consolidate them to a ReadOctetsThatMightBePrimitiveOrConstructed function?